builder/build_validate.go

```diff
@@ -0,0 +1,911 @@
+			return nil, err
+		}
+
+		err = info.getSubPkgs(m.Name, mInfo, pInfo)
+		if err != nil {
+			return nil, err
+		}
+
+		err = info.getSubPkgFiles(m.Name, mInfo, pInfo)
+		if err != nil {
+			return nil, err
+		}
+
+		mcaInfo[m.Name] = info
+	}
+	return mcaInfo, nil
+}
+
+// mcaManInfo collects manifest metadata for each bundle listed in
+// the provided version's MoM.
+func mcaManInfo(config *Builder, version int) (map[string]*swupd.Manifest, error) {
+	mInfo := make(map[string]*swupd.Manifest)
+
+	updateDir := filepath.Join(config.Config.Builder.ServerStateDir, "www")
+	versionDir := filepath.Join(updateDir, fmt.Sprint(version))
+	momPath := filepath.Join(versionDir, "Manifest.MoM")
+
+	mom, err := swupd.ParseManifestFile(momPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get bundle info for each MoM entry
+	for _, f := range mom.Files {
+		// os-core-update-index and iterative manifests are not checked by MCA
+		if f.Name == "os-core-update-index" || f.Type == swupd.TypeIManifest {
+			continue
+		}
+
+		manifestPath := filepath.Join(updateDir, fmt.Sprint(f.Version), "Manifest."+f.Name)
+		manifest, err := swupd.ParseManifestFile(manifestPath)
+		if err != nil {
+			return nil, err
+		}
+
+		bundleInfoPath := filepath.Join(config.Config.Builder.ServerStateDir, "image", fmt.Sprint(version), manifest.Name+"-info")
+		err = manifest.GetBundleInfo(config.Config.Builder.ServerStateDir, bundleInfoPath)
+		if err != nil {
+			return nil, err
+		}
+
+		mInfo[f.Name] = manifest
+	}
+
+	return mInfo, nil
+}
+
+// mcaPkgInfo downloads and queries all packages for each bundle to collect
+// file metadata that will be used to create subtracted package/file lists.
+func mcaPkgInfo(mInfo map[string]*swupd.Manifest, config *Builder, version, downloadRetries int) (map[string]*mcaBundlePkgInfo, error) {
+	// Download RPMs from correct upstream version
+	upstreamVer, err := config.getUpstreamVersion(strconv.Itoa(version))
+	if err != nil {
+		return nil, err
+	}
+
+	bundleDir := filepath.Join(config.Config.Builder.ServerStateDir, "validation")
+	buildVersionDir := filepath.Join(bundleDir, fmt.Sprint(version))
+	packageDir := filepath.Join(buildVersionDir, "packages")
+
+	packagerCmd := []string{
+		"dnf",
+		"-y",
+		"--config=" + config.Config.Builder.DNFConf,
+		"--releasever=" + upstreamVer,
+		"--downloaddir=" + packageDir,
+	}
+
+	// Duplicate package entries can exist across bundles. Use a cache to
+	// avoid re-calculating packages.
+	pkgCache := make(map[string]*pkgInfo)
+
+	pInfo := make(map[string]*mcaBundlePkgInfo)
+
+	// Download and query file metadata from all packages in each bundle
+	for _, m := range mInfo {
+		pInfo[m.Name] = &mcaBundlePkgInfo{
+			allFiles: make(map[string]bool),
+			allPkgs:  make(map[string]*pkgInfo),
+		}
+
+		var pkgList = []string{}
+		for pkg := range m.BundleInfo.AllPackages {
+			pkgList = append(pkgList, pkg)
+		}
+
+		out, err := downloadRpms(packagerCmd, pkgList, buildVersionDir, downloadRetries)
```

Comment:
Packages in the local-rpms folder override the upstream repo, so I plan to disable the local repo for MCA. If the 'Mixer add-rpms' command was able to create a repo that supported release versions, MCA could add support for local RPMs.

Reply:
RPMs that only exist locally would be fine. Would you prefer that I re-enable the local repo and change my warning to identify possible errors with RPM naming conflicts?

Reply:
At what point during clear release is MCA intended to be run? If it was before the files are pushed to the cdn then we might need local repos enabled. If not, I believe it is ok to keep it disable for now if that is the current behavior of the script we are replacing.

Reply:
Yes you are right, I need to re-enable local repos. During the Clear release, it will be run before the files are pushed to the cdn.

Reply:
Thinking about this further, MCA would have issues getting the correct bundle definitions for the release process. Mixer hard codes the upstream bundle definition URL to the clr-bundles repo on Github which will not contain pre-release bundle definitions. This could be fixed by supporting a configurable repo for upstream bundle definitions.

Also, MCA's downstream support for RPM repos and bundle definitions is limited. Neither of these easily supports multiple mix versions.

Reply:
I will add a warning message when local RPMs are enabled.